aead/src/dev.rs

```diff
@@ -1,74 +1,114 @@
+        return Err("decrypt_inout_detached: plaintext mismatch");
+    }
+
+    Ok(())
+}
+
+/// Run AEAD test for the provided failing test vector
+pub fn run_fail_test<C: AeadInOut>(
+    cipher: &C,
+    nonce: &Nonce<C>,
+    aad: &[u8],
+    ct: &[u8],
+) -> Result<(), &'static str> {
+    let res = cipher.decrypt(nonce, Payload { aad, msg: ct });
+    if res.is_ok() {
+        Err("decryption must return error")
+    } else {
+        Ok(())
+    }
+}
 
 /// Define AEAD test
 #[macro_export]
 macro_rules! new_test {
     ($name:ident, $test_name:expr, $cipher:ty $(,)?) => {
         #[test]
         fn $name() {
-            use aead::{
-                Aead, KeyInit, Payload,
-                array::{Array, typenum::Unsigned},
-                dev::blobby::Blob6Iterator,
-            };
-
-            fn run_test(
-                key: &[u8],
-                nonce: &[u8],
-                aad: &[u8],
-                pt: &[u8],
-                ct: &[u8],
-                pass: bool,
-            ) -> Result<(), &'static str> {
-                let key = key.try_into().map_err(|_| "wrong key size")?;
-                let cipher = <$cipher>::new(key);
-                let nonce = nonce.try_into().map_err(|_| "wrong nonce size")?;
-
-                if !pass {
-                    let res = cipher.decrypt(nonce, Payload { aad: aad, msg: ct });
-                    if res.is_ok() {
-                        return Err("decryption must return error");
-                    }
-                    return Ok(());
-                }
-
-                let res = cipher
-                    .encrypt(nonce, Payload { aad: aad, msg: pt })
-                    .map_err(|_| "encryption failure")?;
-                if res != ct {
-                    return Err("encrypted data is different from target ciphertext");
-                }
-                let res = cipher
-                    .decrypt(nonce, Payload { aad: aad, msg: ct })
-                    .map_err(|_| "decryption failure")?;
-                if res != pt {
-                    return Err("decrypted data is different from target plaintext");
-                }
-                Ok(())
-            }
+            use $crate::dev::blobby::Blob6Iterator;
 
             let data = include_bytes!(concat!("data/", $test_name, ".blb"));
             for (i, row) in Blob6Iterator::new(data).unwrap().enumerate() {
                 let [key, nonce, aad, pt, ct, status] = row.unwrap();
-                let pass = match status[0] {
-                    0 => false,
-                    1 => true,
+                let key = key.try_into().expect("wrong key size");
+                let nonce = nonce.try_into().expect("wrong nonce size");
+                let cipher = <$cipher>::new(key);
+
+                let res = match status {
+                    [0] => $crate::dev::run_fail_test(&cipher, nonce, aad, ct),
+                    [1] => $crate::dev::run_pass_test(&cipher, nonce, aad, pt, ct),
```

Comment:
Note that the failing test ignores plaintext. Maybe it's worth to split macros for successful and failing tests?